assets/js/modules/adsense/util/data-mock.js

```diff
@@ -0,0 +1,221 @@
+ */
+function generateDimensionValues( date, dimensions ) {
+	const values = [];
+
+	for ( const dimension of dimensions ) {
+		switch ( dimension.toUpperCase() ) {
+			case 'DATE':
+				values.push( date );
+				break;
+		}
+	}
+
+	return values;
+}
+
+/**
+ * Generates mock data for AdSense reports.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Object} args Report options.
+ * @return {Array.<Object>} An array with generated report.
+ */
+export function getAdSenseMockResponse( args ) {
+	const originalSeedValue = faker.seedValue;
+	const argsHash = parseInt(
+		md5( JSON.stringify( args ) ).substring( 0, 10 ),
+		16,
+	);
+
+	// We set seed for every data mock to make sure that the same arguments get the same report data.
+	// It means that everyone will have the same report data and will see the same widgets in the storybook.
+	// This approach gives us additional flexibility to control randomness on a per widget basis.
+	if ( ! Number.isNaN( argsHash ) ) {
+		faker.seed( argsHash );
+	}
+
+	const metrics = ( Array.isArray( args.metrics ) ? args.metrics : [ args.metrics ] ).filter( ( metric ) => !! metric );
+	const dimensions = ( Array.isArray( args.dimensions ) ? args.dimensions : [ args.dimensions ] ).filter( ( dimension ) => !! dimension );
+
+	const data = {
+		kind: 'adsense#report',
+		startDate: args.startDate,
+		endDate: args.endDate,
+		totalMatchedRows: '0',
+		headers: generateHeaders( metrics, dimensions ),
+		totals: [],
+		averages: [],
+		rows: [],
+	};
+
+	const startDate = new Date( args.startDate );
+	const endDate = new Date( args.endDate );
+	const dayInMilliseconds = 24 * 60 * 60 * 1000;
+	const totalDays = 1 + ( ( endDate - startDate ) / dayInMilliseconds ); // +1 to include the endDate into the dates range.
+
+	// This is the list of operations that we will apply to the range (array) of numbers.
+	const ops = [
+		// Converts range number to a date string.
+		map( ( item ) => ( new Date( startDate.getTime() + ( dayInMilliseconds * item ) ) ).toISOString().split( 'T' )[ 0 ] ),
+		// Add dimension and metric values.
+		map( ( date ) => [ ...generateDimensionValues( date, dimensions ), ...generateMetricValues( date, metrics ) ] ),
+		// Accumulate all rows into a single array.
+		reduce( ( rows, row ) => [ ...rows, row ], [] ),
+	];
+
+	// Process the stream of dates and add generated rows to the report data object.
+	range( 0, totalDays )
+		.pipe( ...ops )
+		.subscribe( ( rows ) => {
+			data.rows = rows;
+			data.totalMatchedRows = rows.length.toString();
+
+			// We pretend that the first row contains averages and the last one totals because we don't
+			// really need mathematically correct values and can simplify the process of finding this information.
+			data.totals = [ ...( rows[ rows.length - 1 ] || [] ) ];
+			data.averages = [ ...( rows[ 0 ] || [] ) ];
+		} );
+
+	// Set the original seed value for the faker.
+	faker.seed( originalSeedValue );
+
+	return data;
+}
```

Comment:
Can we also add a `provideAdSenseMockReport` function here like we do for Analytics? That way we only need a single function to generate and receive the report which is what we'll need in most cases.

Reply:
Added.